### ISSUE TITLE
Support system-wide enforcement of default authentication with proxy

### ISSUE THREAD

**Issue by kedia990** [dotnet/runtime#66587]
## Objective

As an administrator, I'd like to be able to enforce the use of default authentication with the configured proxy for all .NET Core applications that run on systems across the enterprise.

## Current State

Currently, we're able to enforce the use of a proxy, with the `HTTP_PROXY` / `HTTPS_PROXY` / `ALL_PROXY` environment variables. We can set these via Group Policy (or any other mechanism) at the machine level, and rely on .NET Core to use that proxy. Unfortunately, none of those help us with the use of default credentials for authentication (AFAIK).

With .NET Framework, we can achieve our objective by injecting something like this in the system-wide `machine.config` files, which then applies to all .NET Framework-based applications.

```xml
<system.net>
  <defaultProxy useDefaultCredentials="true" />
</system.net>
```

.NET Core also supports a programmatic approach to enforcing default authentication with proxies, like so (done in PowerShell here, but you get the idea):

```powershell
PS C:\> [System.Net.Http.HttpClient]::DefaultProxy.Credentials = [System.Net.CredentialCache]::DefaultCredentials
```

However, we're not in a position to enforce that centrally for any and all .NET Core-based applications. If I've missed a trick that allows us to administratively enforce default credentials for proxies within the current capabilities of .NET Core, please let me know.

## Proposal

I'd propose that we consider introducing another environment variable that, if set, causes the .NET runtime to internally arm the default proxy's credentials with `CredentialCache.DefaultCredentials`. I was able to draft a patch locally to `System.Net.Http` (for .NET 6) and make this work; it could look something like this:

```powershell
# Without environment variable set
PS C:\> $env:PROXY_USE_DEFAULT_CREDENTIALS
PS C:\> Invoke-WebRequest -Uri 'www.google.com'
# Fails with 407 Proxy Authentication Required

# With environment variable set
PS C:\> $env:PROXY_USE_DEFAULT_CREDENTIALS
1
PS C:\> Invoke-WebRequest -Uri 'www.google.com'

StatusCode        : 200
StatusDescription : OK
...
```

If this sounds like a reasonable approach, and you're willing to take a PR for this, I'm happy to try contributing.

**Comment by karelz** [comment]
Triage: Needs more investigation:
- Is `DefaultCredentials` available only on Windows?
- The env. var. scheme comes from curl -- is there something like that there? Or some other SW?

**Comment by kedia990** [comment]
> Is DefaultCredentials available only on Windows?

I haven't explicitly tried this out on Linux (I could, but it'll take me a bit to set up the environment and validate), but according to [this comment](https://github.com/dotnet/runtime/issues/29748#issuecomment-498510691), `DefaultCredentials` should work on Linux under the right circumstances (i.e., if joined to a Kerberos realm).

> The env. var. scheme comes from curl -- is there something like that there? Or some other SW?

Good question. While a bit of searching didn't reveal any interesting examples from other software; for Curl in particular, I was able to find a couple of references ([one](https://stackoverflow.com/questions/45587200/curl-options-equivalent-to-usedefaultcredentials-by-net-webclient), [two](https://curl.se/mail/lib-2012-02/0089.html)). Apparently, if your Curl build has SSPI enabled, you can set `CURLOPT_PROXYUSERPWD` to `:`, and it'll use the default credentials for proxy authentication.

I guess that gives us another path here - we already support extracting credentials from the proxy URL passed in via the existing env vars ([here](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs#L61)). If we didn't want to introduce another environment variable, we could make it so that if the `UserInfo` part of the URL is just `:` (e.g. `http://:@my.cool.proxy:1234`), we could resort to using `DefaultCredentials`. Not necessarily advocating for this, just thinking out loud.

**Comment by wfurt** [comment]
Storing password in ENV is going to be problematic. The `DefaultCredentials` is really empty stub just like `:` curl use. 
If that is sufficient, it should work - NTLM & Kerberos on Windows and Kerberos on Unix platforms.

**Comment by karelz** [comment]
Triage: We are ok with treating empty credentials as default credentials. Not a priority.

**Comment by kedia990** [comment]
Thanks @wfurt, @karelz. Just so I'm clear, by this:

> [@karelz] We are ok with treating empty credentials as default credentials.

You mean,

> we could make it so that if the UserInfo part of the [proxy, via env var] URL is just `:` (e.g. `http://:@my.cool.proxy:1234`), we could resort to using DefaultCredentials.

Is my understanding correct? And if so, would you consider taking a PR for it?

**Comment by wfurt** [comment]
> > we could make it so that if the UserInfo part of the [proxy, via env var] URL is just `:` (e.g. `http://:@my.cool.proxy:1234`), we could resort to using DefaultCredentials.
> 
> Is my understanding correct? And if so, would you consider taking a PR for it?

yes & yes. PR would be welcomed. The change should hopefully not be too difficult. I'm not sure if there is good way how to test it. There are some existing proxy tests or we may extend some of the enterprise  one.
